contributors/design-proposals/federated-statefulsets.md

```diff
@@ -0,0 +1,249 @@
+
+1 – An unique, consistent and discoverable identity for each replica/instance across the federated clusters.
+
+2 – It must be possible for the federated statefulset to first safely form an initial quorum (probably using cluster local replicas and identities).
+It should further also be possible to add members (probably from other clusters and global identities) to this initial quorum.
+
+3 – The ability to scale, across clusters in some deterministic fashion.
+
+4 – An extended design requirement (which might not be a necessary requirement for this phase for work) can be an ability to migrate a pod from one node (in one cluster) to another node (possibly in another cluster), with stable network identity along with its attached volumes.
+
+# Design Proposals
+
+The Statefulset object and APIs are already defined in k8s.
+The same needs to be exposed from federation to ensure that the interfaces remain consistent with a stand-alone k8s cluster.
+A separate controller at federation would however be useful.
+
+We considered two alternative designs for this proposal, both of which are further described in this document.
+We find the Design alternative 2, better in many aspects, and find it more suitable to be pursued for implementation further.
+
+## Design Alternative 1
+
+The federated replicset  controller should be able to spin off and monitor individual pods into federated clusters directly in a similar way to how an in-cluster controller would spin off and monitor the pods in a stand-alone cluster.
+The federation controller should be able to calculate and assign ordinal identities to these pods.
+The federation controller should also enforce the creation sequence, the same way an in-cluster controller would.
+
+### Replica distribution (across federated clusters)
+
+The federated controller will interact with each cluster and create and monitor individual pod objects in the clusters.
+The pod&#39;s spec and state will be stored in individual cluster&#39;s etcd, as it would happen on creating a pod in the given cluster.
+The controller also assigns each ordinal replica to a specific cluster.
+As the primary intention of this feature is to provide a solution for high availability, it makes great sense to ensure that the instances are distributed into maximum clusters possible.
+The initial proposal is to simply hash the replica number into the number of healthy clusters.
+A simple modulo calculation can be used to do the distribution.
+
+### Instance identity and discovery
+
+In the case of in-cluster statefulset, pods discover each other using the in-cluster dns names.
+A headless service with selectors enables creation of dns records against the pod names.
+This cannot work across clusters, as local pod IPs cannot be reached across the clusters.
+
+The proposal to make it work across the clusters is to assign a service type &#39;LoadBalancer&#39; for each pod instance that is created in the clusters.
```

Comment:
Is there any way to do this differently?  Think about this at scale.  Cost, quotas and provisioning would get seriously fun :)  How are we doing guaranteed network identity for each ELB?